File: gsf_checker.py
import requests
import json
import argparse
import sys

URL = 'https://safebrowsing.googleapis.com/v4/threatMatches:find?key='
FILENAME = '/tmp/check.txt'

def get_url(file):
    """
    return safebrowsing url with key;
    accept path to key file
    """
    try:
        with open(file, 'r') as f:
            return URL + f.readlines()[0].rstrip()
    except IOError as err:
        print("Failed with error: %s" % err)
        sys.exit(1)

def get_json(url, data):
    json_data = json.dumps(data)
    headers = {
        'Content-Type': 'application/json'
    }
    try:
        r = requests.post(url, json_data, headers=headers, timeout=5)
    except requests.exceptions.RequestException as err:
        print("Request error: %s" % err)
        sys.exit(1)
    except requests.exceptions.Timeout as err:
        print("Request timeout 5 seconds with error: %s" % err)
        sys.exit(1)
    return r.json()

data = {
        "client": {
                    "clientId":      "myproject",
                    "clientVersion": "1.5.2"
                },
        "threatInfo": {
                    "threatTypes":      ["MALWARE", "SOCIAL_ENGINEERING", "POTENTIALLY_HARMFUL_APPLICATION", "UNWANTED_SOFTWARE"],
                    "platformTypes":    ["ANY_PLATFORM"],
                    "threatEntryTypes": ["URL"],
                    "threatEntries": [
                                {"url": "http://malware.testing.google.test/testing/malware/"},
                                {"url": "http://ianfette.org"},
                            ]
                }
}

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Google Safe browser checker')
    parser.add_argument('--key', '-k', required=True, help='specify the path to  key file')
    path_to_key = parser.parse_args().key

    output = get_json(get_url(path_to_key), data)
    #print(output)

    if len(output) > 0:
        list_data =  output.get('matches')
        with open(FILENAME, 'a') as f:
            for elem in list_data:
                f.writelines(elem.get('threat').get('url') + '\n')
    else:
        open(FILENAME, 'w').close()
        #print "OK"

File: gsf_checker.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"time"
)

type Threat struct {
	Url string `json:"url"`
}

type APIResponse struct {
	Matches []Matches `json:"matches"`
}
type Matches struct {
	ThreatType      string  `json:"threatType"`
	PlatformType    string  `json:"platformType"`
	Threat          *Threat `json:"threat"`
	CacheDuration   string  `json:"cacheDuration"`
	ThreatEntryType string  `json:"threatEntryType"`
}

func getJsonOfRequest(filePath string) []byte {
	data, err := ioutil.ReadFile(filePath)
	if err != nil {
		fmt.Printf("Error open json file: %s", err)
		os.Exit(1)
	}
	return data
}
func getURL(filePath, shortRequestURL string) string {
	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	gsfKey := lines[0]
	return shortRequestURL + gsfKey
}

func sendRequest(requestURL string, jsonOfRequest []byte) []byte {
	//var jsonStr = jsonOfRequest
	req, err := http.NewRequest("POST", requestURL, bytes.NewBuffer(jsonOfRequest))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 4 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()
	//fmt.Println("response status: ", resp.Status)
	//fmt.Println("response header:", resp.Header)

	body, _ := ioutil.ReadAll(resp.Body)
	//fmt.Println("response body:", string(body))
	return body
}

func parseJson(responseData []byte) []string {
	response := bytes.NewReader(responseData)
	decoder := json.NewDecoder(response)
	val := &APIResponse{}
	err := decoder.Decode(val)
	if err != nil {
		log.Fatal(err)
	}
	var matches []string
	if val.Matches != nil {
		for _, s := range val.Matches {
			matches = append(matches, s.Threat.Url)
			//fmt.Println(s.Threat.Url)
		}
	} else {
		fmt.Println("Url not found")
		os.Exit(1)
	}
	return matches
}

func writeLines(urlArray []string, outFile string) {
	file, err := os.Create(outFile)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	for _, line := range urlArray {
		//writer.WriteString(line)
		fmt.Fprintln(writer, line)
	}
	writer.Flush()
}
func main() {
	const URL string = "https://safebrowsing.googleapis.com/v4/threatMatches:find?key="

	keyPath := flag.String("key", "", "usage --key /api.key")
	jsonOfRequestPath := flag.String("json", "", "usage --json /request_data.json")
	outLog := flag.String("out", "", "usage --out /matches_thret_url_to_file")
	flag.Parse()

	key := *keyPath
	requestData := *jsonOfRequestPath
	outSnmpFile := *outLog
	// check required flag
	if key == "" || requestData == "" || outSnmpFile == "" {
		flag.Usage()
		os.Exit(1)
	}

	threatUrls := parseJson(sendRequest(getURL(key, URL), getJsonOfRequest(requestData)))
	//sendRequest(getURL(key, URL))
	writeLines(threatUrls, outSnmpFile)

}
